### PR TITLE
Code for Paper "Bridging the Gap Between Anchor-based and Anchor-free…

### PR DESCRIPTION
**Pull request by hust-kevin**

### CODE REVIEW

**Comment by hust-kevin** [comment]
we re-implement the atss, more detail refer to [office repo](https://github.com/sfzhang15/ATSS) and [paper](https://arxiv.org/abs/1912.02424)


**Re-implement Note**

the core of this paper is `Adaptive Training Sample Selection`, while the final model (coco ap 39.2) actually consists of the following parts.

`ATSS (final model) = RetinaNet(#A=1) + RetinaNet tricks + Imprs on FCOS + atss`

- `RetinaNet(#A=1)` means only one square anchor box per location in RetinaNet
- `RetinaNet tricks` is some trick apply in RetinaNet which boost ap from 35.6 to 36.3, this trick include `use std [0.1, 0.1, 0.2, 0.2] and classification loss normalization`,  in the paper, this is used as the baseline for RetinaNet. more detail refer to [maskrcnn-benckmark](https://github.com/facebookresearch/maskrcnn-benchmark/pull/102). (PS: we don't do experiment on mmdection whether it can get the same boost)
- `Imprs on FCOS` is some trick used in new version of fcos to further impore the result, which include `use GN, GIOU Loss, Centerness` etc
- `atss` is the core part of this paper, which is a bbox assigner mechanism to replace the origin max-iou assigner, more detail refer to paper and code in `core/bbox/assigners/atss_assigner.py`

File: configs/atss/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  year    =  {2019}
+}
+```
+
+
+## Results and Models
+
+| Backbone  | Style   | Lr schd | Mem (GB) | Train time (s/iter) | Inf time (fps) | box AP |
+|:---------:|:-------:|:-------:|:--------:|:-------------------:|:--------------:|:------:|
+| R-50      | pytorch | 1x      |          |                     |                |  38.8  |
```
Comment (yhcao6) [inline configs/atss/README.md]: The official box AP is 39.3, 0.5 AP difference is a large gap. It is necessary to find out where the difference comes.

File: mmdet/core/anchor/atss_target.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
+import torch
```
Comment (yhcao6) [inline mmdet/core/anchor/atss_target.py]: We prefer to put `atss_target`as a function of `atss_head` instead separating it as an independent file.
You can refer the code design of `fcos_head.py`

File: mmdet/core/anchor/atss_target.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        labels = unmap(labels, num_total_anchors, inside_flags)
+        label_weights = unmap(label_weights, num_total_anchors, inside_flags)
+        bbox_targets = unmap(bbox_targets, num_total_anchors, inside_flags)
+        bbox_weights = unmap(bbox_weights, num_total_anchors, inside_flags)
+
+    return (anchors, labels, label_weights, bbox_targets, bbox_weights,
+            pos_inds, neg_inds)
+
+
+def get_num_level_anchors_inside(num_level_anchors, inside_flags):
```
Comment (yhcao6) [inline mmdet/core/anchor/atss_target.py]: We can use `torch.split` to simplify the code.

File: mmdet/core/bbox/assigners/atss_assigner.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+               gt_bboxes_ignore=None,
+               gt_labels=None):
+        """Assign gt to bboxes.
+
+        The assignment is done in following steps
+
+        1. compute iou between all bbox and gt
+        2. compute center distance between all bbox and gt
+        3. on each pyramid level, for each gt, select k bbox whose center
+           are closest to the gt center, so we total select k*l bbox as
```
Comment (yhcao6) [inline mmdet/core/bbox/assigners/atss_assigner.py]: It is better to explain `l` here indicates the number of pyramid levels.

File: mmdet/core/bbox/assigners/atss_assigner.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            num_level_bboxes (List): num of bboxes in each level
+            gt_bboxes (Tensor): Groundtruth boxes, shape (k, 4).
+            gt_bboxes_ignore (Tensor, optional): Ground truth bboxes that are
+                labelled as `ignored`, e.g., crowd boxes in COCO.
+            gt_labels (Tensor, optional): Label of gt_bboxes, shape (k, ).
+
+        Returns:
+            :obj:`AssignResult`: The assign result.
+        """
+        if bboxes.shape[0] == 0 or gt_bboxes.shape[0] == 0:
```
Comment (yhcao6) [inline mmdet/core/bbox/assigners/atss_assigner.py]: Currently we support image contain zero GTs in other assigners.
So it may be better if `atss_assigner` also supports this corner case.

File: mmdet/core/bbox/assigners/atss_assigner.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+            raise ValueError('No gt or bboxes')
+
+        INF = 100000000
+        num_gt = gt_bboxes.size(0)
+        bboxes = bboxes[:, :4]
+        overlaps = bbox_overlaps(bboxes, gt_bboxes)
+
```
Comment (yhcao6) [inline mmdet/core/bbox/assigners/atss_assigner.py]: `atss_assigner` it the core of atss and somewhat complex.
Please add more comments to each part to make it clear and easy to follow.

File: mmdet/models/anchor_heads/atss_head.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+from ..registry import HEADS
+from ..utils import ConvModule, Scale, bias_init_with_prob
+from .anchor_head import AnchorHead
+
+
+@HEADS.register_module
+class ATSSHead(AnchorHead):
+    """
+    https://arxiv.org/abs/1912.02424
+    """
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/atss_head.py]: doc should be more detailed, please refer to the doc of FCOSHead.

File: mmdet/models/anchor_heads/atss_head.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+
+    def __init__(self,
+                 num_classes,
+                 in_channels,
+                 stacked_convs=4,
+                 octave_base_scale=4,
+                 scales_per_octave=1,
+                 conv_cfg=None,
+                 norm_cfg=dict(type='GN', num_groups=32, requires_grad=True),
+                 loss_bbox_weight=2.0,
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/atss_head.py]: `loss_bbox_weight` is redundant with `loss_weight` in `loss_bbox`.

File: mmdet/models/anchor_heads/atss_head.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+        centerness = centerness.permute(0, 2, 3, 1).reshape(-1)
+        bbox_targets = bbox_targets.reshape(-1, 4)
+        labels = labels.reshape(-1)
+        label_weights = label_weights.reshape(-1)
+
+        # classification loss
+        loss_cls = self.loss_cls(
+            cls_score, labels, label_weights, avg_factor=num_total_samples)
+
+        pos_inds = torch.nonzero(labels > 0).squeeze(1)
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/atss_head.py]: It seems that `labels > 0` is redundant.
`pos_inds = torch.nonzero(labels).squeeze(1)` shoud be equiavent to that.

File: mmdet/models/anchor_heads/atss_head.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+                cls_scores,
+                bbox_preds,
+                centernesses,
+                labels_list,
+                label_weights_list,
+                bbox_targets_list,
+                num_total_samples=num_total_samples,
+                cfg=cfg)
+
+        bbox_avg_factor = torch.tensor(bbox_avg_factor).sum()
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/atss_head.py]: `bbox_avg_factor = sum(bbox_avg_factor)`

File: mmdet/models/anchor_heads/atss_head.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+                                      1).reshape(-1, self.cls_out_channels)
+        bbox_pred = bbox_pred.permute(0, 2, 3, 1).reshape(-1, 4)
+        centerness = centerness.permute(0, 2, 3, 1).reshape(-1)
+        bbox_targets = bbox_targets.reshape(-1, 4)
+        labels = labels.reshape(-1)
+        label_weights = label_weights.reshape(-1)
+
+        # classification loss
+        loss_cls = self.loss_cls(
+            cls_score, labels, label_weights, avg_factor=num_total_samples)
```
Comment (yhcao6) [inline mmdet/models/anchor_heads/atss_head.py]: In the official implementation, `num_total_samples` is all_reduced over all gpus and then get the mean.
We may need to check whether it will influence the mAP.

File: mmdet/models/losses/iou_loss.py [hunk 12]
```diff
@@ -69,6 +69,60 @@ def bounded_iou_loss(pred, target, beta=0.2, eps=1e-3):
     return loss
 
 
+@weighted_loss
+def giou_loss(pred, target, eps=1e-6):
+    """GIoU loss.
+
+    https://github.com/sfzhang15/ATSS/blob/master/atss_core/modeling/rpn/atss/loss.py#L36
```
Comment (yhcao6) [inline mmdet/models/losses/iou_loss.py]: Generalized Intersection over Union: A Metric and A Loss for Bounding Box Regression

https://arxiv.org/abs/1902.09630

File: mmdet/models/losses/iou_loss.py [hunk 13]
```diff
@@ -69,6 +69,60 @@ def bounded_iou_loss(pred, target, beta=0.2, eps=1e-3):
...
+        Tensor: Loss tensor.
+    """
+
+    pred_x1 = pred[:, 0]
+    pred_y1 = pred[:, 1]
+    pred_x2 = pred[:, 2]
+    pred_y2 = pred[:, 3]
+    pred_x2 = torch.max(pred_x1, pred_x2)
+    pred_y2 = torch.max(pred_y1, pred_y2)
+    pred_area = (pred_x2 - pred_x1) * (pred_y2 - pred_y1)
```
Comment (yhcao6) [inline mmdet/models/losses/iou_loss.py]: This is a convention in mmdetection. For two coordinates x1, x2, the real length should be x2 - x1 + 1.
You can refer [`bbox_overlaps`](https://github.com/open-mmlab/mmdetection/blob/e907139cb8f4c0195abcf31101925819f369ecc1/mmdet/core/bbox/geometry.py#L63) function.

File: configs/atss/README.md (hunk 1 above)
Comment (hust-kevin) [reply configs/atss/README.md]: we have changed the code according to the `Changes requested` and retrain the model(we train the model use 2 GPUs, to speed up training we set `imgs_per_gpu=4` and  `lr = 0.01 / 2`), now the ap is 39.4, here is the [train log](https://drive.google.com/open?id=1KlPSss2OKpfmWVmKWPxtmKnT0fnGYlFT)



we doubt the difference may caused by one of following reasons

- whether use `all reduce` to compute `num_total_samples` and `bbox_avg_factor`
- when we  train the old model, our server power off during 6th epoch, so we resume training from 5th epoch. theoretically there is no problem, but when we compared the ap of old and retrained model at each epoch, we find that there is a large gap after 5th epoch, and for old model, the ap of 6th epoch is equal ap of 5th epoch, so we doubt that there may be something error when we resume training 

| model | epoch_1 | epoch_2 | epoch_3 | epoch_4 | epoch_5 | epoch_6 | epoch_7 | epoch_8 | epoch_9 | epoch_10 | epoch_11 | epoch_12|
|:-----:|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|:-------:|:--------:|:--------:|:------:|
| `old `|  0.216  |  0.255  |  0.286  |  0.296  |  0.310  |  0.310  |  0.317  |  0.326  |  0.379  |  0.383   |  0.385   |  0.388   |
| `new` |  0.213  |  0.257  |  0.287  |  0.303  |  0.309  |  0.326  |  0.327  |  0.331  |  0.386  |  0.388   |  0.391   |  0.394   |

**Comment by hust-kevin** [comment]
@yhcao6  we re-construct and re-train the model, now ap is 39.4

**Comment by yhcao6** [comment]
> @yhcao6 we re-construct and re-train the model, now ap is 39.4

It is awesome! I would like to add on more that in my experiment, `+1` in GIoU loss also boost the performance. 
I am benchmarking the model on 8gpus since almost all models are trained on 8gpus. I will inform you if there is new progress.
Thanks for you quick reply.

**Comment by openmmlab-v** [comment]
Hi @hust-kevin ！First of all, we want to express our gratitude for your significant PR in the mmdetection project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
